index.d.ts

```diff
@@ -0,0 +1,33 @@
+interface ClearablePromise<T> extends Promise<T> {
+	/**
+	 * Clears the delay and settles the promise.
+	 *
+	 * @memberof ClearablePromise
+	 */
+	clear(): void;
+}
+
+declare const delay: {
+	/**
+	 * Create a promise which resolves after the specified `milliseconds`.
+	 * Optionally pass a `value` to resolve.
+	 *
+	 * @param {number} milliseconds Milliseconds to delay the promise.
+	 * @param {T} [value] Value to resolve in the returned promise.
+	 * @returns {ClearablePromise<T>} a promise which resolves after the specified `milliseconds`
+	 */
+	<T = never>(milliseconds: number, value?: T): ClearablePromise<T>;
```

Comment:
I think you can write the 2 methods separately. This way you have 2 overrides.

```ts
(milliseconds: number): ClearablePromise<void>;
<T = any>(milliseconds: number, value: T): ClearablePromise<T>
```

Reply:
Is there any reason to do that?

Reply:
If you use the two lines, it clearly shows that it has 2 overloads. So you can also show other descriptions per method.

<img width="510" alt="screen shot 2018-08-20 at 12 57 15" src="https://user-images.githubusercontent.com/1913805/44336740-b1c04a00-a478-11e8-9b30-cb3c9f1a5711.png">
<img width="547" alt="screen shot 2018-08-20 at 12 57 21" src="https://user-images.githubusercontent.com/1913805/44336741-b1c04a00-a478-11e8-9d5b-b47a99799bef.png">

If you only have one, it is shown like this
<img width="537" alt="screen shot 2018-08-20 at 12 56 58" src="https://user-images.githubusercontent.com/1913805/44336750-b8e75800-a478-11e8-8ba6-b77f100e5813.png">

Reply:
Sorry, I still don't get the difference.
As for me, the suggested overload is just another way to make the `value` argument optional. The return types would still be the same. Descriptions also will be identical, since it has to be explicitly stated that the method accepts the second argument.
Finally, the suggested overload won't affect intellisense or type inference in any way.

Reply:
> As for me, the suggested overload is just another way to make the value argument optional.

It is just another way of making it optional.

> The return types would still be the same.

The one is returning `ClearablePromise<void>`, the other is returning `ClearablePromise<T>`. With the current implementation, you can write the following line of code.

```ts
const result = await delay<string>(2000);
```

And TypeScript will think that `result` is of type string, which it clearly isn't because there is no `value` provided.

When using overloads, you won't be able to write that line because it isn't accepted by the compiler.

Reply:
Now I see your point, thanks.